textdb/textdb-dataflow/src/main/java/edu/uci/ics/textdb/dataflow/keywordmatch/KeywordMatcher.java

```diff
@@ -0,0 +1,193 @@
+                schema = sourceTuple.getSchema();
+                spanSchema = createSpanSchema();
+            }
+
+
+            for(int attributeIndex = 0; attributeIndex < attributeList.size(); attributeIndex++){
+                IField field = sourceTuple.getField(attributeList.get(attributeIndex).getFieldName());
+                String fieldValue = (String) (field).getValue();
+                if(field instanceof StringField){
+                    //Keyword should match fieldValue entirely
+
+                    if(fieldValue.equals(queryValue.toLowerCase())){
+                        spanIndexValue = 0;
+                        positionIndex = queryValue.length();
+                        addSpanToSpanList(fieldName, spanIndexValue, positionIndex, queryValue, fieldValue);
+                        foundFlag = true;
+
+                    }
+                }
+                else if(field instanceof TextField) {
+                    for (int iter = 0; iter < queryValueArray.size(); iter++) {
+                        String query = queryValueArray.get(iter);
+                        Pattern p = patternList.get(iter);
+                        matcher = p.matcher(fieldValue.toLowerCase());
+                        while (matcher.find(positionIndex) != false) {
+                            spanIndexValue = matcher.start();
+                            positionIndex = spanIndexValue + query.length();
+                            documentValue = fieldValue.substring(spanIndexValue, positionIndex);
+                            addSpanToSpanList(fieldName, spanIndexValue, positionIndex, query, documentValue);
+                            foundFlag = true;
+
+                        }
+                    }
+                }
+                positionIndex = 0;
+            }
+
+            //If all the 'attributes to be searched' have been processed return the result tuple with span info
+            if (foundFlag){
+                foundFlag = false;
+                positionIndex = 0;
+                return getSpanTuple();
+            }
+            //Search next document if the required predicate did not match previous document
+            else if(sourceTuple != null) {
+                positionIndex = 0;
+                spanList.clear();
+
+                return getNextTuple();
+
+            }
+
+            return null;
+
+        } catch (Exception e) {
+            e.printStackTrace();
+            throw new DataFlowException(e.getMessage(), e);
+        }
+
+    }
+
+    private ITuple getSpanTuple() {
+        IField spanListField = new ListField<Span>(new ArrayList<>(spanList));
+        List<IField> fieldListDuplicate = new ArrayList<>(fieldList);
+        fieldListDuplicate.add(spanListField);
+
+        IField[] fieldsDuplicate = fieldListDuplicate.toArray(new IField[fieldListDuplicate.size()]);
+        return new DataTuple(spanSchema, fieldsDuplicate);
+    }
+
+    private Schema createSpanSchema() {
```

Comment:
The Regex team is also writing a similar function.  See https://github.com/TextDB/textdb/pull/84/files 

@laisycs @sandeepreddy602 : can you chime in to see how to avoid duplicate efforts?